### DISCUSSION TITLE
[Lesson 9] setUp failed after 'forge test' in the process of 'deployer key'

### DISCUSSION THREAD

**Discussion by LareineHan** [Cyfrin/foundry-full-course-cu#931] [Q&A] [answered]
Hi, I am getting test failure while adding 'deployer key' during Lesson 9. 

I got lots of test fails, which is different from Patrick's version of fails. So I don't know which I have messed up (I thought everything I followed him correctly..) :(

I run 
```zsh
forge test --fork-url $SEPOLIA_URL 
```
however, I have failed 5 tests and succeed 5 tests, and kept following lessons and did 
```zsh
forge test
``` 
and got this error. 
```zsh
No files changed, compilation skipped
2023-09-27T16:13:18.554330Z ERROR forge::runner: setUp failed reason="Custom Error d8a3fb52:(0x1804c8AB1F12E6bbf3894d4083f33e07309d1f38)" contract=0x34a1d3fff3958843c43ad80f30b94c510645c316

Running 1 test for test/unit/RaffleTest.t.sol:RaffleTest
[FAIL. Reason: Setup failed: Custom Error d8a3fb52:(0x1804c8AB1F12E6bbf3894d4083f33e07309d1f38)] setUp() (gas: 0)
Logs:
  Creating subscription on ChinId:  31337
  Your Subscription Id is:  1
  Please update your Subscription Id in HelperConfig.s.sol
  Funding Subscription  1
  Using vrfCoordinator  0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496
  On ChainID  31337
  Adding Consumer to Raffle  0x62c20Aa1e0272312BC100b4e23B4DC1Ed96dD7D1
  Using vrfCoordinator  0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496
  On ChainID  31337

Test result: FAILED. 0 passed; 1 failed; 0 skipped; finished in 4.29ms
 
Ran 1 test suites: 0 tests passed, 1 failed, 0 skipped (1 total tests)

Failing tests:
Encountered 1 failing test in test/unit/RaffleTest.t.sol:RaffleTest
[FAIL. Reason: Setup failed: Custom Error d8a3fb52:(0x1804c8AB1F12E6bbf3894d4083f33e07309d1f38)] setUp() (gas: 0)

Encountered a total of 1 failing tests, 0 tests succeeded
```
I don't know why I got my Subscription 1 because I added my subscription id on getSepoliaEthConfig()'s returning NetworkConfig. It has 4 digit number got from chainlink. 

Also I set "subscriptionId:0" on getOrCreateAnvilEthConfig()'s  NetworkConfig. So I wonder what the "Your Subscription Id is:  1" means. It's from this function of Interactions.s.sol. 
```solidity 
    function createSubscription(
        address vrfCoordinator
    ) public returns (uint64) {
        console.log("Creating subscription on ChinId: ", block.chainid);
        vm.startBroadcast();
        uint64 subId = VRFCoordinatorV2Mock(vrfCoordinator)
            .createSubscription();
        vm.stopBroadcast();
        console.log("Your Subscription Id is: ", subId);
        console.log("Please update your Subscription Id in HelperConfig.s.sol");
        return subId;
    }
```
Here is my [repo](https://github.com/LareineHan/foundry-smart-contract-lottery-f23.git). 
Please help, amazing community. Thank you in advance!!

**Comment by shikhar229169** [comment, marked as answer]
Actually the create subscription and add consumer function should be called by same person. But over here as you are passing a custom private key as argument in ```vm.startBroadcast()``` while adding a consumer but while creating subscription no argument was passed so it picks it fixed key, as a result of which the txn is performed on behalf of diff users.
So, while creating subscription you need to pass that ```deployerKey``` as argument and problem will be solved and don't forget to add that key in vm.startBroadcast for creating subscription.

**Reply by LareineHan** [reply]
Thanks @shikhar229169. After adding 'deployerKey' , 'forget test' succeed. 
but I am wonder why still my subscription id is '1'. 
```zsh
[PASS] testCheckUpkeepReturnsFalseIfItEnoughTimeHasntPassed() (gas: 19164)
Logs:
  Creating subscription on ChinId:  31337
  Your Subscription Id is:  1
  Please update your Subscription Id in HelperConfig.s.sol
  Funding Subscription  1
  Using vrfCoordinator  0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496
  On ChainID  31337
  Adding Consumer to Raffle  0x50EEf481cae4250d252Ae577A09bF514f224C6C4
  Using vrfCoordinator  0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496
  On ChainID  31337
```

also, after this when I run
```zsh
forge test --fork-url $SEPOLIA_RPC_URL
```

it says 
```zsh
error: a value is required for '--fork-url <URL>' but none was supplied
```

I don't know why it suddenly says none was supplied.

**Reply by shikhar229169** [reply]
- Yes the sub id would be 1 because you ran it on the **local network**, as a result of which from the helper config file you will get sub id as 0 (which indicates a subscription is to be created). So, in the deployer file a new subscription was created and as it was the first one so it was 1.
The thing is whenever you run forge test, it creates a new network chain so the sub id would always be 1 (for the local network).

- Whenever you open a new terminal, you need to load the key-value pairs from the .env file by running ```source .env```

**Reply by LareineHan** [reply]
Thank for the reply! I've added SEPOLIA_RPC_URL indeed in my .env file and runt `source .env` and `echo $SEPOLIA_RPC_URL` prints out a url starts with 'https://eth-sepolia.g.alchemy.com/v2/ with XDFAV)_WEF)E_F stuff'  on my terminal. 
Still I don't know why `none was supplied` ..

**Reply by shikhar229169** [reply]
Can you share the ss for both echo command and test command.
Don't forget to hide the api key in the rpc url.

**Reply by LareineHan** [reply]
Ut-oh, sorry, it's running the test now (maybe I did something very wrong before.. weird) and showing a different error. 
<img width="705" alt="Screenshot 2023-09-27 at 10 37 41 PM" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/128317012/9ff3f7ca-a681-4f4e-a23e-22c6100d817b">

**Reply by shikhar229169** [reply]
I cloned your repo and all tests are passing. Have you modified your tests?

**Reply by LareineHan** [reply]
Yes, I added deployerKey here and there. This is the error deatil of `forge test --fork-url $SEPOLIA_RPC_URL -vvv`
<img width="683" alt="Screenshot 2023-09-27 at 11 30 39 PM" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/128317012/3967af40-2842-488e-8e03-eebe8f6453bf">

**Reply by LareineHan** [reply]
Hey actually now I added `deployerKey` here and there more and finally passed all tests. Now I can catch up the lesson. 
Thanks for helping me @shikhar229169 ! It helped 👍
<img width="675" alt="Screenshot 2023-09-28 at 12 26 04 AM" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/128317012/51589fa1-4b84-4728-b63f-de1a74baac37">

**Reply by EthanHuntIMF99** [reply]
I am getting the same error can you please tell me what you did?? Please attach a screenshot too ;)

**Reply by LareineHan** [reply]
Hi! I added many deployerKey to here and there. I believe some of `deployerKey` adding process was cut or omitted on Patrick's video. Would you check my [repo](https://github.com/LareineHan/foundry-smart-contract-lottery-f23.git) and compare it to yours? I didn't go much from this(your) moment. So you can compare yours with mine. 

Also, good thing I realized is that (you probably know it already) we need to run `source .env` before run fork test. 
I thought it's enough to run it once, but at some moment it lost it's connection or for some reasons when I run `echo $SEPOLIA_RPC_URL` it don't read anything. So make sure before you run fork test, run `source .env`.

**Reply by EthanHuntIMF99** [reply]
I don't understand what you mean by here and there can you please elaborate? Thanx for reaching out to me

**Reply by Merulez** [reply]
Bro i literally copied your code and i still get this error :D , i did the source .env i mean i dont know where is the fck problem .....xD
